kubemarine/core/defaults.py

```diff
@@ -135,29 +135,10 @@ def apply_registry(inventory: dict, cluster: KubernetesCluster) -> dict:
         if inventory['registry'].get('webserver', False):
             thirdparties_address = f"{protocol}://{inventory['registry']['address']}"
 
-    # Patch kubeadm imageRepository
-    if not inventory['services']['kubeadm'].get('imageRepository'):
+    # Patch kubeadm imageRepository and plugin_defaults registry
+    if cluster.raw_inventory.get('services', {}).get('kubeadm', {}).get('imageRepository') is None:
         inventory['services']['kubeadm']["imageRepository"] = registry_mirror_address
-
-    # it is necessary to convert URIs from quay.io/xxx:v1 to example.com:XXXX/xxx:v1
-    if inventory.get('plugin_defaults') is None:
-        inventory['plugin_defaults'] = {}
-    if inventory['plugin_defaults'].get('installation') is None:
-        inventory['plugin_defaults']['installation'] = {}
-    if inventory['plugin_defaults']['installation'].get('registry') is None:
-        inventory['plugin_defaults']['installation']['registry'] = registry_mirror_address
-
-    # The following section rewrites DEFAULT plugins registries and do not touches user-defined registries in plugins
-    # This section required, because plugins defaults contains default non-docker registries and method
-    # "kubemarine.core.defaults.recursive_apply_defaults" will not overwrite this default registries, because it can not
-    # distinguish default from user-defined.
-    # Also, this part of code supports plugin_defaults inventory section and applies everything in accordance with the
-    # priority of the registries.
-    for plugin_name, plugin_params in cluster.inventory['plugins'].items():
-        if cluster.inventory['plugins'][plugin_name].get('installation') is None:
-            cluster.inventory['plugins'][plugin_name]['installation'] = {}
-        if cluster.raw_inventory.get('plugins', {}).get(plugin_name, {}).get('installation', {}).get('registry') is None:
-            cluster.inventory['plugins'][plugin_name]['installation']['registry'] = inventory['plugin_defaults']['installation']['registry']
+    inventory['plugin_defaults']['installation'].setdefault('registry', registry_mirror_address)
```

Comment:
I slightly changed algorithm of registry enrichment. Previously presence of `registry` triggered enrichment of `plugin_defaults`, `plugins.installation.registry`. Default registry `k8s.gcr.io` for nginx-ingress was replaced, and registry for other plugins was enriched. If `registry` was absent but `plugin_defaults` was still present, `plugins.installation.registry` was enriched by different algorithm inside `defaults.apply_defaults` (not `defaults.apply_registry`. In this case registry was other plugins was enriched, but default registry `k8s.gcr.io` for nginx-ingress was **not replaced**.

Now, `registry` triggers enrichment only of CRI configuration, `plugin_defaults`, and `kubeadm.imageRepository`, and thirdparties sources. `plugin_defaults` triggers enrichment of `plugins.installation.registry` independently, as much as `kubeadm.imageRepository` triggers `kubeadm.dns.imageRepository` independently.

I did not write patch for these changes, because consider such configurations that may cause different enrichment result as highly unlikely.